pkg_defs/ska3-core/combine_arch.py

```diff
@@ -41,6 +42,29 @@ def main():
         with open(filename) as fh:
             environments[platform] = {p['name']: p for p in json.load(fh)}
 
+    subtract_environments = {}
+    print(f'Reading environments to subtract:')
+    for env in args.subtract_env:
+        try:
+            platform, filename = env.split('=')
+        except ValueError:
+            print(f' - skipped {env}')
+            continue
+        print(f' + {platform}: {filename}')
+        with open(filename) as fh:
+            subtract_environments[platform] = {p['name']: p for p in json.load(fh)}
+
+    for platform in environments:
+        remove_keys = []
+        for package in environments[platform]:
+            if (platform in subtract_environments
+                    and package in subtract_environments[platform]
+                    and subtract_environments[platform][package]['version']
+                    == environments[platform][package]['version']):
```

Comment:
This can obviously be cleaned up, but this is just the way my mind was working and seemed passable for a utility script.

Reply:
I think the idea of subtracting environments is nice, but does it produce a working package? This procedure produces the complement of ska3-core, but if the intersection of the packages is non-empty, then this will not be a complete package.

I think a requirement is that packages must work. In other words, installing a package should lead to an environment with all the intended tools. The only options then are:
- do this only with packages that do not intersect (do these two intersect?),
- duplicate the intersection of the packages (I am ok with this one),
- make one depend on the other.

Reply:
I don't really understand the question... but part of the problem is likely that I also didn't understand your plan on or how I was supposed to use the current code to make a ska3-perl.  So, to quickly make a ska3-perl that included the package either brought in by ska3-perl-latest or its dependencies, I didn't:

conda create -n ska3-core-flight ska3-core ska3-flight --override-channels -c https://icxc.harvard.edu/aspect/ska3-conda/shiny

and from that make ska3-core-flight-linux.json and ska3-core-flight-osx.json

conda create -n ska3-core-flight-perl ska3-core ska3-flight ska3-perl-latest --override-channels -c https://icxc.harvard.edu/aspect/ska3-conda/shiny

and from that make ska3-core-flight-perl-linux.json and ska3-core-flight-perl-osx.json

And then used the tool like 

combine_arch.py --env linux=ska3-core-flight-perl-linux.json --env osx=ska3-core-flight-perl-osx.json 
  --subtract-env linux=ska3-core-flight-linux.json --subtract-env osx=ska3-core-flight-osx.json  --out meta.yaml

And then rename the meta.yaml package name because I didn't fiddle with the part writing out ska3-core.  So this kind of subtraction seems like it would be useful to make ska3-flight if you didn't already have the packages mechanism... and it *seems* to have put together a reasonable list of packages for the ska3-perl package in this PR.  But I haven't built and tested the package yet.

Reply:
Sorry, I should just have done it. What I meant is to do:

on linux:
```
conda create -n ska3-perl --override-channels -c https://icxc.cfa.harvard.edu/aspect/ska3-conda/shiny ska3-perl-latest
conda activate ska3-perl
conda list --json > ska3-perl-linux.json
```

on OSX
```
conda create -n ska3-perl --override-channels -c https://icxc.cfa.harvard.edu/aspect/ska3-conda/shiny ska3-perl-latest
conda activate ska3-perl
conda list --json > ska3-perl-osx.json
```

then
```
./combine_arch.py --name ska3-perl --version 2020.13+shiny --env linux=ska3-perl-linux.json --env osx=ska3-perl-osx.json --out ska3-perl/meta.yaml
```

That produced this:
```
    - _libgcc_mutex ==0.1  # [linux]
    - _sysroot_linux-64_curr_repodata_hack ==3  # [linux]
    - blas ==1.0  # [linux]
    - bzip2 ==1.0.8  # [linux]
    - ca-certificates ==2020.6.24
    - certifi ==2020.6.20
    - cfitsio ==3.470  # [linux]
    - expat ==2.2.9
    - fontconfig ==2.13.0  # [linux]
    - freetype ==2.10.2  # [linux]
    - gsl ==2.4  # [linux]
    - icu ==58.2  # [linux]
    - intel-openmp ==2020.1  # [linux]
    - kernel-headers_linux-64 ==3.10.0  # [linux]
    - krb5 ==1.18.2  # [linux]
    - ld_impl_linux-64 ==2.33.1  # [linux]
    - libcurl ==7.71.1  # [linux]
    - libcxx ==10.0.0  # [osx]
    - libedit ==3.1.20191231
    - libffi ==3.3
    - libgcc-ng ==9.1.0  # [linux]
    - libgfortran-ng ==7.3.0  # [linux]
    - libiconv ==1.16  # [osx]
    - libpng ==1.6.37  # [linux]
    - libssh2 ==1.9.0  # [linux]
    - libstdcxx-ng ==9.1.0  # [linux]
    - libuuid ==1.0.3  # [linux]
    - libx11-common-cos7-x86_64 ==1.6.7  # [linux]
    - libx11-cos7-x86_64 ==1.6.7  # [linux]
    - libx11-devel-cos7-x86_64 ==1.6.7  # [linux]
    - libxcb ==1.14  # [linux]
    - libxml2 ==2.9.10  # [linux]
    - mkl ==2020.1  # [linux]
    - mkl-service ==2.3.0  # [linux]
    - mkl_fft ==1.1.0  # [linux]
    - mkl_random ==1.1.1  # [linux]
    - ncurses ==6.2
    - numpy ==1.18.5  # [linux]
    - numpy-base ==1.18.5  # [linux]
    - openssl ==1.1.1g
    - perl ==5.26.2
    - perl-app-cpanminus ==1.7044
    - perl-app-env-ascds ==0.04  # [linux]
    - perl-chandra-time ==0.9.2  # [linux]
    - perl-core-deps ==0.3
    - perl-cxc-sysarch ==1.00  # [linux]
    - perl-dbd-sybase ==1.15  # [linux]
    - perl-extended-deps ==0.1  # [linux]
    - perl-ska-agasc ==3.5.0  # [linux]
    - perl-ska-classic ==0.6
    - perl-ska-convert ==4.3  # [linux]
    - perl-ska-web ==4.0  # [linux]
    - perl-tk ==804.035  # [linux]
    - pip ==20.1.1
    - pkg-config ==0.29.2
    - python ==3.8.3
    - readline ==8.0
    - setuptools ==49.2.0
    - six ==1.15.0  # [linux]
    - sqlite ==3.32.3
    - sysroot_linux-64 ==2.17  # [linux]
    - tk ==8.6.10
    - wheel ==0.34.2
    - xorg-kbproto ==1.0.7  # [linux]
    - xorg-libx11 ==1.6.9  # [linux]
    - xorg-libxft ==2.3.2  # [linux]
    - xorg-libxrender ==0.9.10  # [linux]
    - xorg-renderproto ==0.11.1  # [linux]
    - xorg-xproto ==7.0.31  # [linux]
    - xtime ==1.2.2  # [linux]
    - xz ==5.2.5
    - yaml ==0.2.5
    - zip ==3.0
    - zlib ==1.2.11
```

Reply:
Right.  I think a bunch of those packages are in ska3-core so I'm not thrilled with the overlap.

Reply:
I am sorry to keep saying this, but is ska3-core a requirement of ska3-perl? If ska3-perl is intended to always be used with ska3-core, then let's add the explicit requirement.

Reply:
if it is not, then those packages need to be there, and it means that ska3-perl should be generated whenever ska3-core is generated, to make sure there are no conflicts. I am ok with that, since we have a script to do it.

Reply:
Indeed, in this PR I've ignored the fact that ska3-core is really a run requirement for ska3-perl.  

For our use cases (ska3-perl will only ever be installed with ska3-core and ska3-flight) I didn't really see the driver.  If we add ska3-core to the ska3-perl metapackage it just means we'd need to build new ska3-perl metapackages every time we updated ska3-core as you note.  I thought we could get away with fewer package builds.